server/src/main/java/org/opensearch/search/pipeline/StatefulSearchResponseProcessor.java

```diff
@@ -0,0 +1,26 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.pipeline;
+
+import org.opensearch.action.search.SearchRequest;
+import org.opensearch.action.search.SearchResponse;
+
+/**
+ * A specialization of {@link SearchResponseProcessor} that makes use of the request-scoped processor state.
+ * Implementors must implement the processResponse method that accepts request-scoped processor state.
+ */
+public interface StatefulSearchResponseProcessor extends SearchResponseProcessor {
+    @Override
+    default SearchResponse processResponse(SearchRequest request, SearchResponse response) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    SearchResponse processResponse(SearchRequest request, SearchResponse response, PipelinedRequestContext requestContext) throws Exception;
```

Comment:
Should we have dedicated `PipelinedResponseContext responseContext` ? I think it would make sense to separate the request / response state passing.

Reply:
The original motivation for this task was specifically to propagate state from request processors to response processors, like in the oversample/truncate processors in this PR. 

My original thought was to define a new kind of processor (a "bracket" processor) that would implement both the request and response processor interfaces (specifically to accomplish stuff like oversampling on request and undoing that oversampling on response). 

Of course, since processors had already been established to be stateless (and thread-safe for use across multiple requests), that design started to become really awkward. So, when @navneet1v suggested treating it as "passing state through the pipeline", I jumped on it.

So, I specifically want the same context to flow through the whole pipeline, from request processing through to response processing.

Reply:
Got it, so basically there is nothing really request-specific here, but more like: here is the context, may be request processor set something or may be response processor did, we don't really care. May be we could go with  more generic name of `PipelinedProcessingContext` context` ?

Reply:
Sure, that could work.

I'd like a name that makes it very clear that it's scoped to the lifetime of one request/response flow (i.e. it's not persistent).

Reply:
I ended up going with `PipelineProcessingContext` 👍